Flaky test: TestValidateCertificate/invalid_with_disallowed_'CertificateSecretTemplate'_annotations

The tests for Certificate.Spec.SecretTemplate validation may be non-deterministic. They sometimes fail as follows:

```
exec ${PAGER:-/usr/bin/less} "$0" || exit 1
Executing tests from //pkg/internal/apis/certmanager/validation:go_default_test
-----------------------------------------------------------------------------
--- FAIL: TestValidateCertificate (0.00s)
    --- FAIL: TestValidateCertificate/invalid_with_disallowed_'CertificateSecretTemplate'_annotations (0.00s)
        certificate_test.go:747: Expected error spec.secretTemplate.annotations: Invalid value: "cert-manager.io/alt-names": cert-manager.io/* annotations are not allowed but got spec.secretTemplate.annotations: Invalid value: "cert-manager.io/certificate-name": cert-manager.io/* annotations are not allowed
        certificate_test.go:747: Expected error spec.secretTemplate.annotations: Invalid value: "cert-manager.io/certificate-name": cert-manager.io/* annotations are not allowed but got spec.secretTemplate.annotations: Invalid value: "cert-manager.io/alt-names": cert-manager.io/* annotations are not allowed
FAIL
```
-- https://prow.build-infra.jetstack.net/view/gs/jetstack-logs/pr-logs/pull/jetstack_cert-manager/4348/pull-cert-manager-bazel/1426075825404383233

Probably because we iterate over a map here which does not have deterministic order: 
https://github.com/jetstack/cert-manager/blob/9d1c6c9f095a01b4ee563e774b3871810c3934ce/pkg/internal/apis/certmanager/validation/certificate.go#L188-L192

I think it'd make sense to sort the offending annotations and return them in single validation error message.

Alternatively, we could change the tests to check the contents of the expected and actual errors lists and not check that they have the same order. E.g. https://pkg.go.dev/github.com/stretchr/testify/assert#ElementsMatch

/kind bug
/cc @jonathansp

--- a/pkg/internal/apis/certmanager/validation/certificate_test.go
+++ b/pkg/internal/apis/certmanager/validation/certificate_test.go
@@ -18,7 +18,6 @@ package validation
 
 import (
 	"fmt"
-	"reflect"
 	"strings"
 	"testing"
 	"time"
@@ -34,6 +33,7 @@ import (
 	"github.com/jetstack/cert-manager/pkg/internal/api/validation"
 	internalcmapi "github.com/jetstack/cert-manager/pkg/internal/apis/certmanager"
 	cmmeta "github.com/jetstack/cert-manager/pkg/internal/apis/meta"
+	"github.com/stretchr/testify/assert"
 )
 
 var (
@@ -734,25 +734,8 @@ func TestValidateCertificate(t *testing.T) {
 	for n, s := range scenarios {
 		t.Run(n, func(t *testing.T) {
 			errs, warnings := ValidateCertificate(s.a, s.cfg)
-			if len(errs) != len(s.errs) {
-				t.Errorf("Expected errors %v but got %v", s.errs, errs)
-				return
-			}
-			if len(warnings) != len(s.warnings) {
-				t.Errorf("Expected warnings %v but got %v", s.warnings, warnings)
-			}
-			for i, e := range errs {
-				expectedErr := s.errs[i]
-				if !reflect.DeepEqual(e, expectedErr) {
-					t.Errorf("Expected error %v but got %v", expectedErr, e)
-				}
-			}
-			for i, w := range warnings {
-				expectedWarning := s.warnings[i]
-				if w != expectedWarning {
-					t.Errorf("Expected warning %q but got %q", expectedWarning, w)
-				}
-			}
+			assert.ElementsMatch(t, errs, s.errs)
+			assert.ElementsMatch(t, warnings, s.warnings)
 		})
 	}
 }
@@ -864,16 +847,7 @@ func TestValidateDuration(t *testing.T) {
 	for n, s := range scenarios {
 		t.Run(n, func(t *testing.T) {
 			errs := ValidateDuration(&s.cfg.Spec, fldPath)
-			if len(errs) != len(s.errs) {
-				t.Errorf("Expected %v but got %v", s.errs, errs)
-				return
-			}
-			for i, e := range errs {
-				expectedErr := s.errs[i]
-				if !reflect.DeepEqual(e, expectedErr) {
-					t.Errorf("Expected %v but got %v", expectedErr, e)
-				}
-			}
+			assert.ElementsMatch(t, errs, s.errs)
 		})
 	}
 }

